### ISSUE TITLE
False negative when `Self` is used with a callback type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6568]
When a decorator type specifies that the first parameter of the decorated method must be type `MyClass`, and it is applied to an instance method within class `MyClass`, this should result in a type error because the type of `self` in this case is `Self`, not `MyClass`.

```python
from typing import Callable

Callback = Callable[["MyClass", int], str]

def decorator1(method: Callback) -> Callback:
    def wrapper(self: "MyClass", a: int) -> str:
        return "wrapped " + method(self, a)
    return wrapper

class MyClass:
    # This should generate a type error.
    @decorator1
    def method1(self, a: int) -> str:
        return "foo"
```

The proper way to type this is:

```python
S = TypeVar("S", bound="MyClass")

Callback = Callable[[S, int], str]

def decorator1(method: Callback[S]) -> Callback[S]:
    def wrapper(self: S, a: int) -> str:
        return "wrapped " + method(self, a)

    return wrapper
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.338, which I just published. It will also be included in a future release of pylance.

**Comment by mehdigmira** [comment]
@erictraut This new rule logs error for code such as this one

```python
from __future__ import annotations

from collections.abc import Sequence
from typing import Self

from typing_extensions import override


class A:
    def convert(self, items: Sequence[Self])-> Sequence[Self]:
        ...


class B(A):
    @override
    def convert(self, items: Sequence[Self])-> Sequence[Self]:
        return super().convert(get_b())



def get_b() -> list[B]:
    ...
```

I assume the logic behind this is that if I subclass B with a class C, then calling C.convert would be unsafe.

How about allowing this pattern if B is marked as final ? It would allow writing code such as the previous one, which in my opinion is not unsafe really.
